toolchain/check/eval.cpp

```diff
@@ -804,80 +820,53 @@ static auto PerformBuiltinUnaryIntOp(Context& context, SemIRLoc loc,
-  // overflow.
-  switch (builtin_kind) {
-    // Bit shift.
-    case SemIR::BuiltinFunctionKind::IntLeftShift:
-    case SemIR::BuiltinFunctionKind::IntRightShift: {
-      const auto& rhs_orig_val = context.ints().Get(rhs.int_id);
-      if (rhs_orig_val.uge(lhs_val.getBitWidth()) ||
-          (rhs_orig_val.isNegative() && lhs_is_signed)) {
-        CARBON_DIAGNOSTIC(
-            CompileTimeShiftOutOfRange, Error,
-            "shift distance not in range [0, {0}) in {1} {2:<<|>>} {3}",
-            unsigned, TypedInt, BoolAsSelect, TypedInt);
-        context.emitter().Emit(
-            loc, CompileTimeShiftOutOfRange, lhs_val.getBitWidth(),
-            {.type = lhs.type_id, .value = lhs_val},
-            builtin_kind == SemIR::BuiltinFunctionKind::IntLeftShift,
-            {.type = rhs.type_id, .value = rhs_orig_val});
-        // TODO: Is it useful to recover by returning 0 or -1?
-        return SemIR::ErrorInst::SingletonConstantId;
-      }
+namespace {
+// A pair of APInts that are the operands of a binary operator. We use an
+// aggregate rather than `std::pair` to allow RVO of the individual ints.
+struct APIntBinaryOperands {
+  llvm::APInt lhs;
+  llvm::APInt rhs;
+};
+}  // namespace
 
-      if (builtin_kind == SemIR::BuiltinFunctionKind::IntLeftShift) {
-        result_val = lhs_val.shl(rhs_orig_val);
-      } else if (lhs_is_signed) {
-        result_val = lhs_val.ashr(rhs_orig_val);
+// Get a pair of integers at the same suitable bit-width: either their actual
+// width if they have a fixed width, or the smallest canonical width in which
+// they both fit otherwise.
+static auto GetIntsAtSuitableWidth(Context& context, IntId lhs_id, IntId rhs_id,
+                                   IntId bit_width_id) -> APIntBinaryOperands {
+  // Unsized operands: take the wider of the bit widths.
+  if (!bit_width_id.is_valid()) {
+    APIntBinaryOperands result = {.lhs = context.ints().Get(lhs_id),
+                                  .rhs = context.ints().Get(rhs_id)};
+    if (result.lhs.getBitWidth() != result.rhs.getBitWidth()) {
+      if (result.lhs.getBitWidth() > result.rhs.getBitWidth()) {
+        result.rhs = result.rhs.sext(result.lhs.getBitWidth());
       } else {
-        result_val = lhs_val.lshr(rhs_orig_val);
+        result.lhs = result.lhs.sext(result.rhs.getBitWidth());
       }
-      return MakeIntResult(context, lhs.type_id, lhs_is_signed,
-                           std::move(result_val));
     }
-
-    default:
-      // Break to do additional setup for other builtin kinds.
-      break;
+    return result;
   }
 
-  // Other operations are already checked to be homogeneous, so we can extend
-  // the RHS with the LHS bit width.
-  CARBON_CHECK(rhs.type_id == lhs.type_id, "Heterogeneous builtin integer op!");
-  llvm::APInt rhs_val = context.ints().GetAtWidth(rhs.int_id, lhs_bit_width_id);
+  return {.lhs = context.ints().GetAtWidth(lhs_id, bit_width_id),
```

Comment:
The comment on APIntBinaryOperands talks about RVO, but doesn't this function defeat that by returning a named variable in one path and a temporary in another?

Reply:
This `return` statement will have RVO / copy elision applied, regardless of whatever else the function does, because it's directly constructing an instance of the return type in the returned expression.

The other `return` statement will typically have NRVO applied, because all the `return`s in the scope of `result` return `result` (though NRVO is not guaranteed by the language rules).